### ISSUE TITLE
[Backend] S3-02 Maximize XGBoost training accuracy - month-specific features + class weighting + improved labels

### ISSUE THREAD

**Issue by NathanLuckock** [Nathan7108/HackathonPractice#29]
## Problem

The XGBoost training pipeline has four accuracy-limiting issues:

1. **GDELT features are country-level constants.** `_load_country_auxiliary_features()` computes GDELT features once for the entire file. Every training row for Ukraine gets identical GDELT values regardless of month. The model cannot learn temporal GDELT signals.

2. **ACLED 90-day context is missing.** `_acled_features_from_group()` only sees one month of events. `acled_event_count_90d` equals the current month count (wrong), and `acled_event_acceleration` is hardcoded to 1.0.

3. **No class weighting.** Most country-months are LOW. The model over-predicts LOW and under-predicts HIGH/CRITICAL.

4. **Labels use only fatalities.** A month with 0 deaths but 500 protests, 30 battles, and 50 explosions is labeled LOW. Real risk includes event diversity.

## Goal

Maximize XGBoost accuracy by making features time-aware and improving labeling and class balance.

### Fix 1: Month-specific GDELT features

Replace `_load_country_auxiliary_features()` GDELT logic. Instead of computing once per country, compute per training row inside `build_training_dataset()`.

For each ACLED country-month, load the GDELT CSV and filter to a window around that month:

```python
# Inside build_training_dataset(), before the groupby loop:
gdelt_df = None
gdelt_path = root / "data" / "gdelt" / f"{code}_events.csv"
if gdelt_path.exists():
    try:
        gdelt_df = pd.read_csv(gdelt_path)
        gdelt_df.columns = gdelt_df.columns.str.strip()
        gdelt_df["date"] = pd.to_datetime(gdelt_df["SQLDATE"].astype(str), format="%Y%m%d", errors="coerce")
        gdelt_df = gdelt_df.dropna(subset=["date"])
    except Exception:
        gdelt_df = None

# Inside the groupby loop, for each period:
if gdelt_df is not None and not gdelt_df.empty:
    # Get end of this month as reference date
    month_end = period.to_timestamp() + pd.offsets.MonthEnd(0)
    # Filter GDELT to 90 days before month_end
    mask_90 = gdelt_df["date"] > (month_end - pd.Timedelta(days=90))
    mask_30 = gdelt_df["date"] > (month_end - pd.Timedelta(days=30))
    gdelt_window = gdelt_df[mask_90 & (gdelt_df["date"] <= month_end)]
    if not gdelt_window.empty:
        gdelt_features = compute_gdelt_features(gdelt_window, window_days=30)
    else:
        gdelt_features = compute_gdelt_features(pd.DataFrame(), window_days=30)
    features.update(gdelt_features)
```

For weekly aggregate format (BigQuery CSVs with `_weekly_aggregate` column), the date filtering works the same way since SQLDATE is the Monday of each week.

### Fix 2: Rolling ACLED 90-day context

Replace `_acled_features_from_group()` with a version that takes the full country ACLED DataFrame plus the current period, so it can compute proper 90-day windows:

```python
def _acled_features_from_period(full_df, period, window_days=30):
    month_end = period.to_timestamp() + pd.offsets.MonthEnd(0)
    month_start = period.to_timestamp()
    
    recent_30 = full_df[(full_df["event_date"] > month_end - pd.Timedelta(days=30)) & (full_df["event_date"] <= month_end)]
    recent_90 = full_df[(full_df["event_date"] > month_end - pd.Timedelta(days=90)) & (full_df["event_date"] <= month_end)]
    older_60 = full_df[(full_df["event_date"] > month_end - pd.Timedelta(days=90)) & (full_df["event_date"] <= month_end - pd.Timedelta(days=30))]
    
    # Current month group
    current = full_df[(full_df["event_date"] >= month_start) & (full_df["event_date"] <= month_end)]
    
    fatalities = pd.to_numeric(current["fatalities"], errors="coerce").fillna(0)
    total_fatal = float(fatalities.sum())
    event_type = current["event_type"].astype(str)
    
    return {
        "acled_fatalities_30d": total_fatal,
        "acled_battle_count": int((event_type == "Battles").sum()),
        "acled_civilian_violence": int((event_type == "Violence against civilians").sum()),
        "acled_explosion_count": int((event_type == "Explosions/Remote violence").sum()),
        "acled_protest_count": int(event_type.isin(["Protests", "Riots"]).sum()),
        "acled_fatality_rate": total_fatal / max(window_days, 1),
        "acled_event_count_90d": len(recent_90),
        "acled_event_acceleration": float(len(recent_30) / max(len(older_60), 1)),
        "acled_unique_actors": int(current["actor1"].nunique()) if "actor1" in current.columns else 0,
        "acled_geographic_spread": int(current["admin1"].nunique()) if "admin1" in current.columns else 0,
    }
```

### Fix 3: Class weighting

Add `sample_weight` to XGBoost training based on inverse class frequency:

```python
# In train_risk_scorer(), after building X and y:
from collections import Counter
counts = Counter(y)
total = len(y)
n_classes = len(counts)
weights = {cls: total / (n_classes * count) for cls, count in counts.items()}
sample_weights = np.array([weights[label] for label in y])

# Pass to fit:
model.fit(X_train, y_train, sample_weight=sample_weights_train, ...)
```

Also split `sample_weights` using the same train/test split indices.

### Fix 4: Improved labeling

Replace `_label_from_fatalities()` with a composite scoring function:

```python
def _label_from_events(group):
    fatalities = pd.to_numeric(group.get("fatalities", pd.Series(0)), errors="coerce").fillna(0).sum()
    event_type = group.get("event_type", pd.Series(dtype=object)).astype(str)
    battles = (event_type == "Battles").sum()
    explosions = (event_type == "Explosions/Remote violence").sum()
    civilian_violence = (event_type == "Violence against civilians").sum()
    protests = event_type.isin(["Protests", "Riots"]).sum()
    
    # Composite score: fatalities dominate, but event diversity matters
    score = (
        fatalities * 1.0
        + battles * 5.0
        + explosions * 4.0
        + civilian_violence * 3.0
        + protests * 0.2
    )
    
    if score >= 1000:
        return "CRITICAL"
    if score >= 400:
        return "HIGH"
    if score >= 100:
        return "ELEVATED"
    if score >= 20:
        return "MODERATE"
    return "LOW"
```

### In scope
- All four fixes in `backend/ml/risk_scorer.py`
- The `build_training_dataset()` and `train_risk_scorer()` functions
- Import `numpy as np` at top of file (for sample_weight)

### Out of scope
- Changes to `predict_risk()` (inference path unchanged)
- Changes to other ML modules
- LSTM forecaster

## Acceptance Criteria

- [ ] Each training row has month-specific GDELT features (not all identical per country)
- [ ] `acled_event_acceleration` is NOT 1.0 for all rows
- [ ] `acled_event_count_90d` differs from current month event count
- [ ] Classification report shows support for all 5 classes (or 4 if CRITICAL merges into HIGH)
- [ ] Overall accuracy >= 80% on test set
- [ ] Top 10 feature importances include features from at least 3 data sources (ACLED + GDELT + one of UCDP/WB)
- [ ] Training completes without errors in under 30 minutes
- [ ] `predict_risk()` still works after retraining

## Manual Test Steps

1. `python -m backend.ml.data.fetch_all_countries --step models`
2. Check "Dataset shape" — should be similar row count to before
3. Check "Risk label distribution" — should have meaningful counts for MODERATE, ELEVATED, HIGH (not 95% LOW)
4. Check classification report accuracy
5. Check top 10 features — should be diverse across data sources
6. Quick inference test: `python -c "from backend.ml.risk_scorer import predict_risk; print(predict_risk({}))"` — should return a valid prediction

## Cursor Agent Instructions

**Repo:** https://github.com/Nathan7108/HackathonPractice
**Working directory:** C:\Users\natha\HackathonPractice

### Step-by-step Implementation Plan

1. Open `backend/ml/risk_scorer.py`
2. Add `import numpy as np` at top
3. Replace `_label_from_fatalities()` with `_label_from_events(group)` that uses composite scoring (Fix 4)
4. Add `_acled_features_from_period(full_df, period, window_days=30)` function (Fix 2)
5. Rewrite `build_training_dataset()`:
   - For each country, load GDELT CSV once and parse dates
   - Load UCDP and World Bank as country-level features (keep existing `_load_country_auxiliary_features` for these two)
   - In the groupby loop: call `_acled_features_from_period()` instead of `_acled_features_from_group()`
   - In the groupby loop: compute month-specific GDELT features from the windowed GDELT data (Fix 1)
   - Use `_label_from_events(group)` instead of `_label_from_fatalities(fatalities)` (Fix 4)
6. In `train_risk_scorer()`: add class weighting via sample_weight (Fix 3)
   - After `y = le.transform(...)`, compute inverse-frequency weights
   - Split sample_weights alongside X_train/X_test
   - Pass `sample_weight=sample_weights_train` to `model.fit()`
7. Keep `predict_risk()` unchanged
8. Keep `_acled_features_from_group()` (still used elsewhere) but the training loop should use the new `_acled_features_from_period()`

### Important edge cases
- GDELT CSVs may be weekly aggregates (check for `_weekly_aggregate` column). `compute_gdelt_features()` already handles this — just pass the filtered DataFrame
- Some countries have no GDELT data — use empty DataFrame fallback
- `period.to_timestamp()` returns the first day of the month; `+ pd.offsets.MonthEnd(0)` gets the last day
- `sample_weight` array must be split with the same indices as X_train/y_train — use the indices from `train_test_split` with `numpy` indexing

### Validation Checklist
- [ ] No import errors
- [ ] `build_training_dataset()` returns DataFrame with varied GDELT features per row
- [ ] Classification accuracy >= 80%
- [ ] Feature importances span multiple data sources
- [ ] `predict_risk({})` returns valid dict

### Definition of Done
- All 4 fixes implemented, training runs cleanly, accuracy >= 80%, diverse feature importances

### If You Get Stuck
- GDELT date filtering on weekly aggregates: the SQLDATE column is already parseable as `%Y%m%d` (Monday of week). Filtering by date range works identically.
- If `period.to_timestamp()` errors: `pd.Timestamp(str(period))` is an alternative
- If sample_weight causes shape mismatch: ensure you index it with the same train indices from `train_test_split` — use `numpy` array indexing, not pandas
